src/runtime/long.cpp

```diff
@@ -1191,49 +1192,81 @@ Box* longRTrueDiv(BoxedLong* v1, Box* _v2) {
+    return boxFloat(pow_float_float(lhs_float, rhs->d));
+}
+
+Box* longPow(BoxedLong* lhs, Box* rhs, Box* mod) {
+    if (!isSubclass(lhs->cls, long_cls))
+        raiseExcHelper(TypeError, "descriptor '__pow__' requires a 'long' object but received a '%s'",
+                       getTypeName(lhs));
+
+    BoxedLong* mod_long = nullptr;
+    if (mod != None) {
+        if (isSubclass(mod->cls, long_cls)) {
+            mod_long = static_cast<BoxedLong*>(mod);
+        } else if (isSubclass(mod->cls, int_cls)) {
+            mod_long = boxLong(static_cast<BoxedInt*>(mod)->n);
+        } else {
+            return NotImplemented;
         }
-        RELEASE_ASSERT(mpz_sgn(mod->n) >= 0, "");
     }
 
-    if (isSubclass(_v2->cls, long_cls)) {
-        BoxedLong* v2 = static_cast<BoxedLong*>(_v2);
-        BoxedLong* r = new BoxedLong();
-        mpz_init(r->n);
-
-        RELEASE_ASSERT(mpz_sgn(v2->n) >= 0, "");
+    BoxedLong* rhs_long = nullptr;
+    if (isSubclass(rhs->cls, long_cls)) {
+        rhs_long = static_cast<BoxedLong*>(rhs);
+    } else if (isSubclass(rhs->cls, int_cls)) {
+        rhs_long = boxLong(static_cast<BoxedInt*>(rhs)->n);
+    } else {
+        return NotImplemented;
+    }
 
-        if (mod) {
-            mpz_powm(r->n, v1->n, v2->n, mod->n);
+    if (mod != None) {
+        if (isSubclass(mod->cls, float_cls)) {
+            raiseExcHelper(TypeError, "pow() 3rd argument not "
+                                      "allowed unless all arguments are integers");
         } else {
-            RELEASE_ASSERT(mpz_fits_ulong_p(v2->n), "");
-            uint64_t n2 = mpz_get_ui(v2->n);
-            mpz_pow_ui(r->n, v1->n, n2);
+            if (mpz_sgn(rhs_long->n) < 0)
+                raiseExcHelper(ValueError, "pow() 2nd argument "
+                                           "cannot be negative when 3rd argument specified");
+            else if (mpz_sgn(mod_long->n) == 0)
+                raiseExcHelper(ValueError, "pow() 3rd argument cannot be 0");
         }
+    }
+
+    BoxedLong* r = new BoxedLong();
+    mpz_init(r->n);
+
+    if (mpz_sgn(rhs_long->n) == 0 && mod == None) {
+        mpz_set_ui(r->n, 1l);
         return r;
-    } else if (isSubclass(_v2->cls, int_cls)) {
-        BoxedInt* v2 = static_cast<BoxedInt*>(_v2);
-        RELEASE_ASSERT(v2->n >= 0, "");
-        BoxedLong* r = new BoxedLong();
-        mpz_init(r->n);
-        if (mod)
-            mpz_powm_ui(r->n, v1->n, v2->n, mod->n);
-        else
-            mpz_pow_ui(r->n, v1->n, v2->n);
-        return r;
+    } else if (mpz_sgn(rhs_long->n) == -1) {
+        BoxedFloat* rhs_float = static_cast<BoxedFloat*>(longFloat(rhs_long));
+        BoxedFloat* lhs_float = static_cast<BoxedFloat*>(longFloat(lhs));
+        return boxFloat(pow_float_float(lhs_float->d, rhs_float->d));
+    }
+
+    if (mod_long) {
+        mpz_pow_ui(r->n, lhs->n, mpz_get_ui(rhs_long->n));
+        return longMod(r, mod_long);
```

Comment:
I think we need to be careful here if the long doesn't fit in an integer, ie `mpz_get_ui` will return just the low bits of the long.  For 1 and -1 as the base, we have the same behavior as CPython since it's only the lowest bit that matters.  For 0 though, we give the wrong result: we say `0 ** (1<<100)` is 1 (since we treat `1<<100` as 0),, but CPython correctly says it is 0.

For other bases, CPython will crash (will try to do the exponentiation even though there is not enough memory), and we will just silently return the wrong value.  ex `2 ** (1<<100)` for us is 1.

My suggestion is that we should call `mpz_fits_ulong_p`, and if that is true then do this path.  If it doesn't fit, then we check+handle 0/1/-1 cases, and for the others we abort() (not sure it's safe to throw an OverflowError when CPython would have crashed).

For the pow case with a mod, I think it's important that we do the pow+mod as a single step, rather than pow-then-mod -- integrating the two means that the result can be computed even if the pow would not fit in memory.  ex `pow(1<<100, 1<<100, 1000)` works just fine and should return 376.  The modulus-sign rules sound tricky, and I can never remember how python does those, but we should either try to apply them to the result of mpz_powm, or we should abort with an "unimplemented" message :/